File: extended/plugins/send-message-python/send_message_plugin/kubernetes.py
from __future__ import annotations

import base64
import json
import os
import ssl
import urllib.error
import urllib.parse
import urllib.request
from http import HTTPStatus
from typing import Any

from .models import (
    AUTH_HEADER,
    BEARER_PREFIX,
    DEFAULT_KUBERNETES_URL,
    ChannelResource,
    SERVICE_ACCOUNT_CA,
    SERVICE_ACCOUNT_TOKEN,
)
from .util import as_dict, deep_get, string_value


class InClusterKubernetesClient:
    def __init__(self) -> None:
        self.token = SERVICE_ACCOUNT_TOKEN.read_text(encoding="utf-8").strip()
        self.base_url = kubernetes_base_url()
        self.ssl_context = ssl.create_default_context(cafile=str(SERVICE_ACCOUNT_CA))

    def get_message_channel(self, namespace: str, name: str) -> ChannelResource:
        response = self._get_json(
            "/apis/ee.kargo.akuity.io/v1alpha1/namespaces/"
            f"{quote(namespace)}/messagechannels/{quote(name)}"
        )
        return ChannelResource(
            secret_name=string_value(deep_get(response, "spec", "secretRef", "name")),
            slack_channel_id=string_value(deep_get(response, "spec", "slack", "channelID")),
            resource_kind="MessageChannel",
            resource_name=string_value(deep_get(response, "metadata", "name")),
            resource_namespace=string_value(deep_get(response, "metadata", "namespace")),
        )

    def get_cluster_message_channel(self, name: str) -> ChannelResource:
        response = self._get_json(
            "/apis/ee.kargo.akuity.io/v1alpha1/clustermessagechannels/"
            f"{quote(name)}"
        )
        return ChannelResource(
            secret_name=string_value(deep_get(response, "spec", "secretRef", "name")),
            slack_channel_id=string_value(deep_get(response, "spec", "slack", "channelID")),
            resource_kind="ClusterMessageChannel",
            resource_name=string_value(deep_get(response, "metadata", "name")),
        )

    def get_secret(self, namespace: str, name: str) -> dict[str, str]:
        response = self._get_json(
            f"/api/v1/namespaces/{quote(namespace)}/secrets/{quote(name)}"
        )
        decoded: dict[str, str] = {}
        for key, value in as_dict(response.get("data")).items():
            try:
                raw = base64.b64decode(string_value(value))
            except Exception as exc:
                raise RuntimeError(f'error decoding secret key "{key}": {exc}') from exc
            decoded[str(key)] = raw.decode("utf-8")
        return decoded

    def _get_json(self, path: str) -> dict[str, Any]:
        request = urllib.request.Request(
            urllib.parse.urljoin(self.base_url.rstrip("/") + "/", path.lstrip("/")),
            method="GET",
            headers={AUTH_HEADER: f"{BEARER_PREFIX}{self.token}"},
        )
        try:
            with urllib.request.urlopen(
                request,
                context=self.ssl_context,
                timeout=30,
            ) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace").strip()
            if exc.code == HTTPStatus.NOT_FOUND:
                raise RuntimeError(f"resource not found at {path}") from exc
            raise RuntimeError(
                f"Kubernetes API error {exc.code} {exc.reason}: {body}"
            ) from exc


def kubernetes_base_url() -> str:
    host = os.environ.get("KUBERNETES_SERVICE_HOST", "").strip()
    if not host:
        return DEFAULT_KUBERNETES_URL
    port = (
        os.environ.get("KUBERNETES_SERVICE_PORT_HTTPS", "").strip()
        or os.environ.get("KUBERNETES_SERVICE_PORT", "").strip()
        or "443"
    )
    return f"https://{host}:{port}"


def quote(value: str) -> str:
    return urllib.parse.quote(value, safe="")

File: extended/plugins/send-message-ts/src/kubernetes.ts
import { readFileSync } from "node:fs";
import { request as httpsRequest } from "node:https";
import { join } from "node:path";

import {
  apiGroup,
  apiVersion,
  bearerPrefix,
  defaultKubernetesBaseURL,
  defaultServiceAccountDir,
} from "./constants.js";
import { asErrorMessage, asRecord, requiredString } from "./support.js";
import type { ChannelResource, KubernetesClient } from "./types.js";

export class InClusterKubernetesClient implements KubernetesClient {
  private readonly baseURL: URL;
  private readonly ca: string;
  private readonly token: string;

  public constructor(serviceAccountDir = defaultServiceAccountDir) {
    const host = process.env.KUBERNETES_SERVICE_HOST?.trim();
    const port = process.env.KUBERNETES_SERVICE_PORT?.trim();
    const baseURL =
      host && port
        ? `https://${host}:${port}`
        : host
          ? `https://${host}`
          : defaultKubernetesBaseURL;

    this.baseURL = new URL(baseURL);
    this.token = readFileSync(join(serviceAccountDir, "token"), "utf8").trim();
    this.ca = readFileSync(join(serviceAccountDir, "ca.crt"), "utf8");
  }

  public async getMessageChannel(
    namespace: string,
    name: string,
  ): Promise<ChannelResource> {
    const resource = await this.requestResource(
      `/apis/${apiGroup}/${apiVersion}/namespaces/${encodeURIComponent(namespace)}/messagechannels/${encodeURIComponent(name)}`,
      `message channel ${namespace}/${name}`,
    );
    return parseChannelResource(resource, "MessageChannel");
  }

  public async getClusterMessageChannel(name: string): Promise<ChannelResource> {
    const resource = await this.requestResource(
      `/apis/${apiGroup}/${apiVersion}/clustermessagechannels/${encodeURIComponent(name)}`,
      `cluster message channel ${name}`,
    );
    return parseChannelResource(resource, "ClusterMessageChannel");
  }

  public async getSecret(
    namespace: string,
    name: string,
  ): Promise<Record<string, string>> {
    const resource = await this.requestResource(
      `/api/v1/namespaces/${encodeURIComponent(namespace)}/secrets/${encodeURIComponent(name)}`,
      `Secret ${namespace}/${name}`,
    );
    const data = asRecord(resource.data, `Secret ${namespace}/${name}.data`);

    return Object.fromEntries(
      Object.entries(data).map(([key, value]) => {
        if (typeof value !== "string") {
          throw new Error(
            `Secret ${namespace}/${name}.data.${key} must be a base64 string`,
          );
        }
        return [key, Buffer.from(value, "base64").toString("utf8")];
      }),
    );
  }

  private async requestResource(
    path: string,
    displayName: string,
  ): Promise<Record<string, unknown>> {
    const { statusCode, body } = await requestJSON({
      ca: this.ca,
      headers: {
        Authorization: `${bearerPrefix}${this.token}`,
        Accept: "application/json",
      },
      host: this.baseURL.hostname,
      method: "GET",
      path,
      port: this.baseURL.port || "443",
    });

    if (statusCode === 404) {
      throw new Error(`${displayName} not found`);
    }
    if (statusCode >= 400) {
      throw new Error(
        `error reading ${displayName} from Kubernetes API: ${statusCode} ${body}`.trim(),
      );
    }

    try {
      return asRecord(JSON.parse(body), displayName);
    } catch (error) {
      throw new Error(
        `error decoding ${displayName} response: ${asErrorMessage(error)}`,
      );
    }
  }
}

function parseChannelResource(
  resource: Record<string, unknown>,
  fallbackKind: string,
): ChannelResource {
  const metadata = asRecord(resource.metadata, "metadata");
  const spec = asRecord(resource.spec, "spec");
  const secretRef = asRecord(spec.secretRef, "spec.secretRef");
  const slack = asRecord(spec.slack, "spec.slack");

  return {
    secretName: requiredString(secretRef, "name", "spec.secretRef.name"),
    slackChannelID: requiredString(slack, "channelID", "spec.slack.channelID"),
    resourceKind:
      typeof resource.kind === "string" ? resource.kind : fallbackKind,
    resourceName: requiredString(metadata, "name", "metadata.name"),
    resourceNamespace:
      typeof metadata.namespace === "string" ? metadata.namespace : undefined,
  };
}

async function requestJSON(options: {
  ca: string;
  headers: Record<string, string>;
  host: string;
  method: string;
  path: string;
  port: string;
}): Promise<{ body: string; statusCode: number }> {
  return new Promise((resolve, reject) => {
    const request = httpsRequest(
      {
        ca: options.ca,
        headers: options.headers,
        host: options.host,
        method: options.method,
        path: options.path,
        port: options.port,
      },
      (response) => {
        const chunks: Buffer[] = [];
        response.on("data", (chunk) => {
          chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });
        response.on("end", () => {
          resolve({
            body: Buffer.concat(chunks).toString("utf8"),
            statusCode: response.statusCode ?? 500,
          });
        });
      },
    );

    request.on("error", reject);
    request.end();
  });
}

